sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -65,35 +89,102 @@ class EquivalentExpressions {
   }
 
   /**
-   * Adds the expression to this data structure recursively. Stops if a matching expression
-   * is found. That is, if `expr` has already been added, its children are not added.
+   * Adds each expression to structural expression data structure, grouping them with existing
+   * structurally equivalent expressions. Non-recursive.
+   */
+  def addStructExpr(ctx: CodegenContext, expr: Expression): Unit = {
+    if (expr.deterministic) {
```

Comment:
We cannot always share a function for non-deterministic cases? e.g.,
```
int subExpr1 = input[0] + random();
int subExpr2 = input[1] + random();
=>
int subExpr1 = subExpr(input[0]);
int subExpr2 = subExpr(input[1]);

int subExpr(int v) { return v + random(); }
```

Reply:
Non-deterministic expressions can't do sub-expression elimination.

Reply:
oh, I see.

Reply:
btw, this idea is limited to common subexprs? For example, the idea can cover a case like;
```
select sum(a + b), sum(b + c), sum(c + d), sum(d + e) from values (1, 1, 1, 1, 1) t(a, b, c, d, e)
```
?

Reply:
It is probably suitable. Only if we want these functions sum(a+b)...etc. to be called in split functions. Their inputs can be parameterized.